cmd/gitserver/server/vcs_syncer_perforce.go

```diff
@@ -235,6 +292,57 @@ func p4ping(ctx context.Context, host, username, password string) error {
 	return nil
 }
 
+// p4depots returns all of the depots to which the user has access on the host
+// and whose names match the given nameFilter, which can contain asterisks (*) for wildcards
+// if nameFilter is blank, return all depots
+func p4depots(ctx context.Context, host, username, password, nameFilter string) ([]PerforceDepot, error) {
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	var cmd *exec.Cmd
+	if nameFilter == "" {
+		cmd = exec.CommandContext(ctx, "p4", "-Mj", "-ztag", "depots")
+	} else {
+		cmd = exec.CommandContext(ctx, "p4", "-Mj", "-ztag", "depots", "-e", nameFilter)
+	}
+	cmd.Env = append(os.Environ(),
+		"P4PORT="+host,
+		"P4USER="+username,
+		"P4PASSWD="+password,
+	)
+
+	out, err := runWith(ctx, wrexec.Wrap(ctx, log.NoOp(), cmd), false, nil)
+	if err != nil {
+		if ctxerr := ctx.Err(); ctxerr != nil {
+			err = ctxerr
+		}
+		if len(out) > 0 {
+			err = errors.Errorf("%s (output follows)\n\n%s", err, specifyCommandInErrorMessage(string(out), cmd))
+		}
+		return nil, err
+	}
+	depots := make([]PerforceDepot, 0)
+	if len(out) > 0 {
+		// the output of `p4 -Mj -ztag depots` is a series of JSON-formatted depot defintions, one per line
+		buf := bufio.NewScanner(bytes.NewBuffer(out))
+		for buf.Scan() {
+			depot := PerforceDepot{}
+			err := json.Unmarshal(buf.Bytes(), &depot)
+			if err != nil {
+				return nil, errors.Errorf("malformed output from p4 depots: %s", err)
```

Comment:
We should wrap errors instead of creating new ones.

```suggestion
				return nil, errors.Wrap(err, "malformed output from p4 depots")
```

Reply:
good idea!